src/Oscoin/Storage/Block/Orphanage.hs

```diff
@@ -270,17 +297,99 @@ storeOrphan
     -> Orphanage c tx s
 storeOrphan b o = o { orphans = Map.insert (blockHash b) b (orphans o) }
 
+{------------------------------------------------------------------------------
+
+                                Pruning chains
+
+ If a suitable 'ChainCandidate' has been found, the winning chain needs to be
+ removed from the 'Orphanage'. We offer two different functions, depending from
+ the context: 'pruneOrphanageShallow' removes only the winning chain, but it
+ doesn't touch any chain branching from any of the blocks of the chain being
+ pruned. If that's not the desired behaviour, you might want to consider using
+ 'pruneOrphanageDeep'. In pictures:
+
+
+          ORIGINAL                   SHALLOW                         DEEP
+
+                 |                         |                            |
+                 |                         |                            |
+                 |                         |                            |
+    Winning      |                         |                            |
+    Candidate    |                         |                            |
+            +-+  |                         |                            |
+            | | /\                         \                            \
+            | |-  -\                        -\                           -\
+            |<|     >                         >                            >
+            |||
+            |||
+            |v|
+            |-|                       /-
+          /-|||                     /-
+        /-  |||                   /-
+      <-    |||                 <-
+            |v|                         \
+            | |\                         \
+            | | v                         v
+            | |
+            +-+
+
+------------------------------------------------------------------------------}
+
+-- | Like 'pruneOrphanageShallow', but also prunes any other candidate chain
+-- branching off from /any/ block of the chain being pruned.
+pruneOrphanageDeep
+    :: (Ord (BlockHash c)
+       , Eq s
+       )
+    => BlockHash c
+    -> ChainCandidate c s
+    -> Orphanage c tx s
+    -> Orphanage c tx s
+pruneOrphanageDeep rootHash bestFork@ChainCandidate{candidateChain} o =
+    foldl' (flip bulkDelete)
```

Comment:
`bulkDelete` does not remove `orphans`. This probably introduces a memory leak. Also, we don’t remove chains forking off forks. I have the feeling that maybe `pruneOrphanageDeep` should call `pruneOrphanageShallow` recursively.

Reply:
Yeah, I do feel the same, but the implementation it's quite head-wrecking so I went with the easiest I could think of, heh.

Reply:
Ok, fine with me for now.